### PR TITLE
CORTX-34012: Query deployment modification with new design

### PR DESCRIPTION
**Pull request by Pranali04796**
Signed-off-by: Pranali Ugale <pranali.ugale@seagate.com>

# Pull Request
## Problem Statement
-   Query deployment modification with new design
-  JIRA : [CORTX-34012](https://jts.seagate.com/browse/CORTX-34012)

## Design
-   For Bug, Describe the fix here.
[Query Deployment REST API specification - Private-Cortx - Confluence (atlassian.net)](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1024229377/Query+Deployment+REST+API+specification)
[FS: CORTX K8s Deployment for LR Reference Config - Private-Cortx - Confluence (atlassian.net)](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1051295778/FS+CORTX+K8s+Deployment+for+LR+Reference+Config)
-   \[ \] Coding conventions are followed and code is consistent

-   \[ \] Unit and System Tests are added
-   \[ \] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   \[ \] Testing was performed with RPM
- Build details
image
- Deployment Details
image

1. GET TOPLOLGY:
`curl --location --request GET 'https://ssc-vm-g4-rhev4-1916.colo.seagate.com:31169/api/v2/system/topology'`

{
    "topology": {
        "cluster_id": "b39c55a2-479e-411a-8cc0-83c5694a6718",
        "version": "2.0.0-7640",
        "nodes": [
            {
                "id": "7ba8990a85d38508ca9bac4ff516d3ea",
                "version": "2.0.0-7640",
                "services": [
                    "io"
                ],
                "type": "data_node/0",
                "storage_set": "storage-set-1",
                "deployment_time": "1661333744",
                "hostname": "cortx-data-g0-0.cortx-data-headless.cortx.svc.cluster.local",
                "storage": [
                    {
                        "devices": {
                            "data": [
                                "/dev/sdd",
                                "/dev/sde"
                            ],
                            "metadata": [
                                "/dev/sdc"
                            ]
                        },
                        "name": "cvg-01",
                        "type": "ios"
                    }
                ]
            },
            {
                "id": "7d9ee991ae6ce628a0e10294bdebaa6b",
                "version": "2.0.0-7640",
                "type": "ha_node",
                "storage_set": "storage-set-1",
                "deployment_time": "1661333716",
                "hostname": "cortx-ha-headless"
            },
            {
                "id": "aa378e768a2ba5291d786b3dc48d5650",
                "version": "2.0.0-7640",
                "services": [
                    "io"
                ],
                "type": "data_node/1",
                "storage_set": "storage-set-1",
                "deployment_time": "1661333749",
                "hostname": "cortx-data-g1-0.cortx-data-headless.cortx.svc.cluster.local",
                "storage": [
                    {
                        "devices": {
                            "data": [
                                "/dev/sdg",
                                "/dev/sdh"
                            ],
                            "metadata": [
                                "/dev/sdf"
                            ]
                        },
                        "name": "cvg-02",
                        "type": "ios"
                    }
                ]
            },
            {
                "id": "ac2f9656e38a92990522025e1f3ccaef",
                "version": "2.0.0-7640",
                "services": [
                    "rgw_s3"
                ],
                "type": "server_node",
                "storage_set": "storage-set-1",
                "deployment_time": "1661333780",
                "hostname": "cortx-server-0.cortx-server-headless.cortx.svc.cluster.local"
            },
            {
                "id": "ecabccb4bced858e282f568d6bf04558",
                "version": "2.0.0-7640",
                "services": [
                    "agent"
                ],
                "type": "control_node",
                "storage_set": "storage-set-1",
                "deployment_time": "1661333725",
                "hostname": "cortx-control"
            }
        ],
        "storage_sets": [
            {
                "id": "storage-set-1",
                "durability": {
                    "data": "1+0+0",
                    "metadata": "1+0+0"
                }
            }
        ],
        "certificates": [
            {
                "subject": {
                    "countryName": "IN",
                    "stateOrProvinceName": "MH",
                    "localityName": "Pune",
                    "organizationName": "Seagate Technology",
                    "commonName": "seagate.com"
                },
                "issuer": {
                    "countryName": "IN",
                    "stateOrProvinceName": "MH",
                    "localityName": "Pune",
                    "organizationName": "Seagate Technology",
                    "commonName": "seagate.com"
                },
                "not_valid_after": "2032-03-05 09:47:00",
                "not_valid_before": "2022-03-08 09:47:00",
                "serial_number": 641940933613535422020029400965858427869994973022,
                "version": "Version.v3",
                "name": "s3.seagate.com.pem",
                "type": "SSL Certificate"
            }
        ]
    }
}
image
image

image
image
image

-   \[ \] Interface change (if any) are documented
-   \[ \] Side effects on other features (deployment/upgrade)
-   \[ \] Dependencies on other component(s)

-   \[ \] JIRA number/GitHub Issue added to PR
-   \[ \] PR is self reviewed
-   \[ \] Jira and state/status is updated and JIRA is updated with PR link
-   \[ \] Check if the description is clear and explained

-   \[ \] Changes done to WIKI / Confluence page / Quick Start Guide
- [Query Deployment REST API specification - Private-Cortx - Confluence (atlassian.net)](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1024229377/Query+Deployment+REST+API+specification)

[FS: CORTX K8s Deployment for LR Reference Config - Private-Cortx - Confluence (atlassian.net)](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/1051295778/FS+CORTX+K8s+Deployment+for+LR+Reference+Config)

### CODE REVIEW

File: csm/conf/etc/csm/csm.conf [hunk 1]
```diff
@@ -191,3 +191,8 @@ STORAGE_CAPACITY_MANAGMENT:
 RETRY:
   retry_count: 5
   sleep_duration: 3
+TOPLOLOGY:
+    backend: "consul"
```
Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/csm.conf:195]: backend -> storage

File: csm/conf/etc/csm/csm.conf [hunk 2]
```diff
@@ -191,3 +191,8 @@ STORAGE_CAPACITY_MANAGMENT:
...
+    index: "conf"
+    object_store: "cortx"
```
Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/csm.conf:197]: object_store -> name. 
cortx -> CORTX

File: csm/core/agent/csm_agent.py [hunk 3]
```diff
@@ -135,6 +136,18 @@ def _configure_s3_services():
         CsmRestApi._app[const.S3_BUCKET_SERVICE] = BucketService(s3_plugin_obj)
         CsmRestApi._app[const.S3_CAPACITY_SERVICE] = S3CapacityService(s3_plugin_obj)
 
+    @staticmethod
+    def get_topology_config():
+        host = Conf.get(const.DATABASE_INDEX, const.CONSUL_CONFIG_HOST)
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:141]: This seems repetitive

Comment (pawankumarsrivastava) [reply csm/core/agent/csm_agent.py:141]: Can we skip get call and use a variable already loaded?

File: csm/core/agent/csm_agent.py [hunk 4]
```diff
@@ -135,6 +136,18 @@ def _configure_s3_services():
...
+        port = Conf.get(const.DATABASE_INDEX, const.DB_CONSUL_CONFIG_PORT)
+        index = Conf.get(const.CSM_GLOBAL_INDEX, const.TOPOLOGY_INDEX)
+        backend = Conf.get(const.CSM_GLOBAL_INDEX, const.TOPOLOGY_BACKEND)
+        topology_config = {
+            const.BACKEND : Conf.get(const.CSM_GLOBAL_INDEX, const.TOPOLOGY_OBJECT_STORE),
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:146]: const.BACKEND --> Change it to same as i mentioned above

File: csm/core/controllers/information.py [hunk 5]
```diff
@@ -77,3 +76,90 @@ async def post(self):
...
+            raise e
+        except Exception as e:
+            Log.error(f'Unable to fetch topology information: {e}')
+            raise CsmInternalError("Unable to fetch topology information.")
+        Log.info(
+            f"Processed request: {self.request.method} {self.request.path}")
+        return CsmResponse(response)
+
+@CsmAuth.public
+@CsmView._app_routes.view("/api/v2/system/topology/{resource}/{resource_id}")
```
Comment (pawankumarsrivastava) [inline csm/core/controllers/information.py:136]: Resource ID will not be present for certificate. Have you handled it?

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/controllers/information.py (hunk 5 above)
Comment (Pranali04796) [reply csm/core/controllers/information.py:136]: yes

File: csm/common/certificate.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    def _get_name_details(self, rdns):
+        """
+        Get x509 Name object details (i.e Subject and Issuer)
+        rdns: Relatively Distinguished Names
+        """
+        name_details = {}
+        for name in rdns:
+            if name._attributes:
+                name_details[name._attributes[0].oid._name] = name._attributes[0].value
+        return name_details
```
Comment (rohitkolapkar) [inline csm/common/certificate.py:85]: Add new line at the EOF

File: csm/conf/etc/csm/csm.conf (hunk 2 above)
Comment (rohitkolapkar) [inline csm/conf/etc/csm/csm.conf:197]: Are you maintaining default_csm.conf  with new additions ?

File: csm/core/controllers/information.py [hunk 7]
```diff
@@ -77,3 +76,90 @@ async def post(self):
...
+    """
+    GET: Complete deployment topology
+    """
+    def __init__(self, request):
+        super().__init__(request)
+        self._service = self.request.app[const.INFORMATION_SERVICE]
+
+    async def get(self):
+        """GET REST implementation for complete topology."""
+        Log.info(f"Processing request: {self.request.method} {self.request.path}")
```
Comment (rohitkolapkar) [inline csm/core/controllers/information.py:92]: add request id in logs

File: csm/core/controllers/information.py [hunk 8]
```diff
@@ -77,3 +76,90 @@ async def post(self):
...
+        self._service = self.request.app[const.INFORMATION_SERVICE]
+
+    async def get(self):
+        """GET REST implementation to query information about resource from deployment topology."""
+        Log.info(f"Processing request: {self.request.method} {self.request.path}")
+        resource = self.request.match_info[const.ARG_RESOURCE]
+        # Check for valid Resource
+        if resource not in const.TOPOLOGY_RESOURCES:
+            raise CsmNotFoundError(f"Invalid resource: {resource}")
+        # Call Information Service
```
Comment (rohitkolapkar) [inline csm/core/controllers/information.py:122]: Remove unnecessary comments

File: csm/core/services/query_deployment/topology_factory.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        except ConfError as e:
+            Log.error(f'Unable to fetch topology information: {e}')
+            raise CsmInternalError("Unable to fetch topology information.")
+        except Exception as e:
+            Log.error(f'Unable to fetch topology information: {e}')
+            raise CsmInternalError("Unable to fetch topology information.")
+        return payload
+
+class TopologyFactory:
+    "The Factory Class"
```
Comment (rohitkolapkar) [inline csm/core/services/query_deployment/topology_factory.py:178]: please update docstring.

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/conf/etc/csm/csm.conf (hunk 1 above)
Comment (Pranali04796) [reply csm/conf/etc/csm/csm.conf]: Removed entry

File: csm/conf/etc/csm/csm.conf (hunk 2 above)
Comment (Pranali04796) [reply csm/conf/etc/csm/csm.conf]: Updated

File: csm/core/agent/csm_agent.py (hunk 4 above)
Comment (Pranali04796) [reply csm/core/agent/csm_agent.py]: Corrected

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/agent/csm_agent.py (hunk 3 above)
Comment (Pranali04796) [reply csm/core/agent/csm_agent.py]: removed get call

File: csm/common/certificate.py (hunk 6 above)
Comment (Pranali04796) [reply csm/common/certificate.py]: added

File: csm/conf/etc/csm/csm.conf (hunk 2 above)
Comment (Pranali04796) [reply csm/conf/etc/csm/csm.conf]: yes added

File: csm/core/controllers/information.py (hunk 7 above)
Comment (Pranali04796) [reply csm/core/controllers/information.py]: Updated

File: csm/core/controllers/information.py (hunk 8 above)
Comment (Pranali04796) [reply csm/core/controllers/information.py]: Removed

File: csm/core/services/query_deployment/topology_factory.py (hunk 9 above)
Comment (Pranali04796) [reply csm/core/services/query_deployment/topology_factory.py]: Updated docstring

File: csm/core/blogic/const.py [hunk 10]
```diff
@@ -1011,3 +1011,33 @@
...
+URL = 'url'
+BACKEND = 'backend'
+TOPOLOGY_DICT_INDEX = "topology_dict_index"
+METADATA = 'metadata'
+DIX = 'dix'
+SNS = 'sns'
+PARITY = 'parity'
+SPARE = 'spare'
+PROVISIONING = 'provisioning'
+DEVICES = 'devices'
```
[the remark this answers is not in the document]
Comment (Pranali04796) [reply csm/core/blogic/const.py]: Added new line

**Review by rohitkolapkar — APPROVED** [review]
LGTM

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by kupranay — APPROVED** [review]

**Pull request merged**